File: ruby/Tree.rb
class BinarySearchTree
  attr_accessor :root
  def initialize(data = nil)
    @root = Node.new(data)
  end

  def add_to_search_tree(node = @root, data)
    return @root = Node.new(data) if node.data.nil?
    if data < node.data
      return node.left = Node.new(data) if node.left.nil?
      return add_to_search_tree(node.left, data) if node.left != nil?
    elsif data > node.data
      return node.right = Node.new(data) if node.right.nil?
      return add_to_search_tree(node.right, data) if !node.right.nil?
    end
    nil
  end

end

File: python/tree.py
from node import Node
class BinarySearchTree:
    def __init__(self, data=None):
        self.root = Node(data)

    def add_to_search_tree(self, node=None, data=None):
        if node is None:
            node = self.root
        if node.data is None:
            node.data = data
            return node
        if data < node.data:
            if node.left is None:
                node.left = Node(data)
                return node.left
            return self.add_to_search_tree(node.left, data)
        elif data > node.data:
            if node.right is None:
                node.right = Node(data)
                return node.right
            return self.add_to_search_tree(node.right, data)
        return None

    def in_order_traversal(self, node=None, result=[]):
        if node is None:
            node = self.root
        if node.left is not None:
            self.in_order_traversal(node.left, result)
        result.append(node.data)
        if node.right is not None:
            self.in_order_traversal(node.right, result)
        return result

    def pre_order_traversal(self, node=None, result=[]):
        if node is None:
            node = self.root
        result.append(node.data)
        if node.left is not None:
            self.pre_order_traversal(node.left, result)
        if node.right is not None:
            self.pre_order_traversal(node.right, result)
        return result
